cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -485,6 +647,36 @@ ArrayKernelExec ArithmeticExecFromOp(detail::GetTypeId get_id) {
   }
 }
 
+template <template <typename... Args> class KernelGenerator, typename Op>
+ArrayKernelExec GenerateArithmeticWithFloatOutType(detail::GetTypeId get_id) {
```

Comment:
Instead, wouldn't it make more sense to just insert an implicit cast to floating point?

Reply:
Used [`ArithmeticFloatingPointFunction`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc#L1151) instead which casts integers to floating-point.